src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpUseAssignableTypeForForeachVariable.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.NetCore.Analyzers.Runtime;
+
+namespace Microsoft.NetCore.CSharp.Analyzers.Runtime
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class CSharpUseAssignableTypeForForeachVariable : UseAssignableTypeForForeachVariable
+    {
+        protected override void AnalyzeLoop(OperationAnalysisContext context, Compilation compilation)
+        {
+            if (context.Operation is not IForEachLoopOperation loop || loop.Syntax is not CommonForEachStatementSyntax syntax)
+            {
+                return;
+            }
+
+            var loopInfo = loop.SemanticModel.GetForEachStatementInfo(syntax);
+            if (!loopInfo.CurrentConversion.Exists ||
+                !loopInfo.CurrentConversion.IsImplicit ||
+                !loopInfo.CurrentConversion.IsIdentity)
+            {
+                return;
+            }
+
+            ITypeSymbol collectionElementType = loopInfo.ElementType;
+            ITypeSymbol variableType = ((IVariableDeclaratorOperation)loop.LoopControlVariable).Symbol.Type;
```

Comment:
Are we sure this cast is safe? Also isn't it possible to end-up with a null symbol which would result in a NRE on the `.Type` call?

Reply:
@Evangelink Good call!

The cast is **not** safe for a case like `foreach (var (a, b) in x)`.

A test demonstrating this should be added.

*found it in roslyn's tests*:

https://github.com/dotnet/roslyn/blob/51718d0c6a6c3adba33efebf815e68f194b82a39/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs#L1101-L1102

That's why I love roslyn being open source ❤️

Reply:
Tuples, right! I couldn't figure out what else could be there except `IVariableDeclaratorOperation`. Thank you =)